### PR TITLE
Allow blacklisting blocks from being finalized again after block revert

### PR DESCRIPTION
**Pull request by sorpaas**
https://github.com/paritytech/substrate/issues/6224, cc @rphmeier
Hopefully I understand the use case.
Still needs to write tests.

This adds functionality to allow blacklisting blocks from being finalized again after call to `revert`. If `blacklist_finalized` is set, reverted finalized blocks will be put into an in-mem blacklist. Upon finalization logic again, the client checks the blacklist and early return if it finds a block in the blacklist. The blacklist currently resets upon client restart.

### CODE REVIEW

**Comment by burdges** [comment]
We must "divert" finality quickly.  After votes though, we could change our minds back and once again permit finality, so how this influences for-choice get messy.  If you've validated the disputed candidate then you'll happily build on this block probably.

We might ask Alistair if node could vote for this in grandpa once they validate the disputed candidate successfully, but maybe they should only vote on chain containing enough validity statements.

**Comment by sorpaas** [comment]
How about we just introduce another type of "finality" (call it "probably finalized")? Keep the semantics of current one as it is, but when doing finalization, we first mark the height as "probably finalized", then after a longer delay, mark it "finalized". This allows us to keep the assumption about state pruning about "finalized", and also maintain usability.

File: client/db/src/lib.rs [hunk 1]
```diff
@@ -1483,7 +1490,12 @@ impl<Block: BlockT> sc_client_api::backend::Backend<Block> for Backend<Block> {
 		})
 	}
 
-	fn revert(&self, n: NumberFor<Block>, revert_finalized: bool) -> ClientResult<NumberFor<Block>> {
+	fn revert(
```
Comment (arkpar) [inline client/db/src/lib.rs:1493]: This function should additionally check that the state we are reverting to is still available (by calling  `self.have_state_at()`)

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -957,6 +958,11 @@ impl<Block: BlockT> Backend<Block> {
 		// TODO: ensure best chain contains this block.
 		let number = *header.number();
 		self.ensure_sequential_finalization(header, last_finalized)?;
+
+		if self.finalize_blacklist.read().contains(hash) {
+			return Ok((*hash, number, false, false))
```
Comment (arkpar) [inline client/db/src/lib.rs:963]: Why not return an error?

**Comment by arkpar** [comment]
Have you considered using `BlockRules` for the blacklist as I was suggesting in https://github.com/paritytech/substrate/issues/6224#issuecomment-638317138 ?

To prevent syncing to blacklisted chains `Client` needs to return `KnownBad` status for blacklisted blocks when `status` or `import` are called.

File: client/db/src/lib.rs (hunk 2 above)
Comment (sorpaas) [reply client/db/src/lib.rs:963]: In this case although we don't finalize the block, we still want to keep importing the block, right?
If we return error here, I think the overall importing would fail.

**Comment by burdges** [comment]
@sorpaas We could eventually devise "finality score" statistics based upon availability coverage, no-shows among approval checkers and even backing checkers, and reports from parties fishermen that we have not yet defined, which should increase your confidence beyond merely grandpa finality.  We've a long ways to go on the theory side before such things make sense though.

**Comment by gavofyork** [comment]
what's the status of this... it's lookign stale? cc @arkpar @sorpaas

File: client/api/src/backend.rs [hunk 3]
```diff
@@ -418,15 +418,28 @@ pub trait Backend<Block: BlockT>: AuxStore + Send + Sync {
 	) -> sp_blockchain::Result<()>;
 
 	/// Commit block insertion.
-	fn commit_operation(&self, transaction: Self::BlockImportOperation) -> sp_blockchain::Result<()>;
+	///
+	/// The `finalization_check` closure checks that if an imported block is able to be
+	/// finalized. If the check returns false, the import succeeds, but with the imported block
+	/// unfinalized.
+	fn commit_operation<F: Fn(&Block::Hash) -> bool>(
```
Comment (arkpar) [inline client/api/src/backend.rs:425]: Why not pass a flag instead of a closure?

File: client/api/src/backend.rs [hunk 4]
```diff
@@ -418,15 +418,28 @@ pub trait Backend<Block: BlockT>: AuxStore + Send + Sync {
...
 
 	/// Finalize block with given Id.
 	///
 	/// This should only be called if the parent of the given block has been finalized.
-	fn finalize_block(
+	///
+	/// The `finalization_check` closure checks that if an imported block is able to be
+	/// finalized. If the check returns false, the import succeeds, but with the imported block
+	/// unfinalized.
+	fn finalize_block<F: Fn(&Block::Hash) -> bool>(
```
Comment (arkpar) [inline client/api/src/backend.rs:438]: It seems like `finalization_check` should really be checked before calling this function. Also documentation mentions "import" which is not part of this function.

File: client/service/src/client/block_rules.rs [hunk 5]
```diff
@@ -30,18 +30,31 @@ use sc_client_api::{ForkBlocks, BadBlocks};
 pub enum LookupResult<B: BlockT> {
 	/// Specification rules do not contain any special rules about this block
 	NotSpecial,
-	/// The bock is known to be bad and should not be imported
+	/// The block is known to be bad and should not be imported
 	KnownBad,
+	/// The block is known not to be finalized
+	KnownUnfinalized,
```
Comment (arkpar) [inline client/service/src/client/block_rules.rs:36]: I'd call it something like `FinalityBlacklist`. "Unfinalized" sounds like it might still be finalized. Also "block that should not be finalized" practically means it should not be imported as well, right @rphmeier ? So from the perspective of `BlockRules` it is the same as `KnownBad`

File: client/api/src/backend.rs (hunk 3 above)
Comment (sorpaas) [reply client/api/src/backend.rs:425]: This is for finalization check -- when `Client` calls it, it needs to access the block rules which is not available in backend. https://github.com/paritytech/substrate/pull/6301/files#diff-d78994a221a27b69ca7b0d31ff112120R246

So I'm not sure if it's possible to only pass a flag here.

Comment (arkpar) [reply client/api/src/backend.rs:425]: `finalization_check` accepts just the block hash. So the client can simply compute the the same closure before calling `commit_operation` and pass the result as a flag. Or, even better, make it a property of `BlockImportOperation`

Comment (sorpaas) [reply client/api/src/backend.rs:425]: The issue is -- in one `BlockImportOperation` there may be multiple finalization attempts. So that `finalization_check` may be called multiple times with different result.

Comment (arkpar) [reply client/api/src/backend.rs:425]: So maybe simply remove the blocks that need not to be finalized from `transaction. finalized_blocks`? The point is that all the data needs to be commited should be in `transaction` already. Introducing additional filters at this stage complicates the interface and makes the logic hard to follow.

File: client/service/src/client/block_rules.rs (hunk 5 above)
Comment (rphmeier) [reply client/service/src/client/block_rules.rs:36]: Yeah, it should not be imported.

Comment (sorpaas) [reply client/service/src/client/block_rules.rs]: Changed this to directly insert it into BadBlock so it won't get imported.

File: client/api/src/backend.rs (hunk 3 above)
Comment (sorpaas) [reply client/api/src/backend.rs]: Now that we directly insert the finalization blacklist to BadBlock, this is not an issue any more!

File: client/api/src/backend.rs [hunk 6]
```diff
@@ -418,7 +418,14 @@ pub trait Backend<Block: BlockT>: AuxStore + Send + Sync {
 	) -> sp_blockchain::Result<()>;
 
 	/// Commit block insertion.
-	fn commit_operation(&self, transaction: Self::BlockImportOperation) -> sp_blockchain::Result<()>;
+	///
+	/// The `finalization_check` closure checks that if an imported block is able to be
+	/// finalized. If the check returns false, the import succeeds, but with the imported block
+	/// unfinalized.
```
Comment (arkpar) [inline client/api/src/backend.rs:424]: Obsolete comment

**Comment by arkpar** [comment]
Looks good, apart from one obsolete comment.
